### PR TITLE
Add --dump flag to build

### PR DESCRIPTION
**Pull request by discordianfish**
When specified, apply will dump all manifests to stdout instead of
applying them.

This is useful to validate changes to kfdef and kubeflow upgrades.

### CODE REVIEW

**Comment by discordianfish** [comment]
Tests fail because:
```
Insert deployment error: googleapi: Error 403: Quota exceeded for quota group 'deploymentMutato
rs' and limit 'Write queries per day' of service 'deploymentmanager.googleapis.com' for consumer 'project_number:29647740582'., rateLimitExceeded
```

**Comment by discordianfish** [comment]
Looks like this isn't working since it breaks the order in which the resources get applied. Looking into that now.

**Comment by discordianfish** [comment]
Ok this should do it!

**Comment by jlewi** [comment]
Thanks @discordianfish 

Is there anyway to achieve this functionality using off the shelf functionality rather than continuing to build out a Kubeflow specific CLI?

IIUC the problem is that we have a directories of independent kustomize packages and we'd like to run `kustomize build` in all those directories.

If I had to stack rank my preferred solutions it would be

1. kustomize
1. kpt
1. kfctl

I think this can be done via a [kpt function](https://googlecontainertools.github.io/kpt/reference/fn/source/) which is now upstreamed into kustomize at least on master.

Here's an example.
https://github.com/kubernetes-sigs/kustomize/blob/266af31bcb7b471e5135ccd8d35b1d24d09158a9/functions/examples/injection-tshirt-sizes/image/main.go

Could we adapt that to

1. Find all kustomization.yaml files
1. Apply the kustomization.yaml and emit the resulting files

**Comment by discordianfish** [comment]
@jlewi But I need this to work with my existing 1.0.1 setup, so I can compare the output to whatever next version we will deploy.

In general yeah, I could do without kfctl in general and only use kubeflow stock tools. But again, whatever they will do, I like to compare the current kfctl generated manifests to it.

I mean, unless there is a better way to validate that an update is safe, given that updating to the next kubeflow release won't be a few line change in my kfdef but more involved.

That being said, I now have this functionality so it doesn't really matter to me whether to merge this.

**Comment by jlewi** [comment]
I think this is valuable and don't want to block a good enough solution.

Should we be using "build" and not "apply" though to generate hydrated manifests? Maybe something like

```
kfctl build -dump
```

**Comment by jlewi** [comment]
@discordianfish at @yanniszark thoughts?

I have a strong preference for this to be a subcommand of build not apply. This seems more consistent with the semantics of kustomize and a shift left philosophy. Under a shift left philosophy build would produce hydrated manifests (which is effectively what this command) is doing and this would then be piped to the apply stage which could be kfctl, kubectl, or other tools like ArgoCD?

**Comment by vpavlin** [comment]
My 2c... `--dry-run` for `apply` does not make much sense semantically since dry running apply would mean to me that you are trying to verify that all the produced resources are applicable in the cluster.

If the goal is to generate all the resources to be able to compare multiple revisions, then the suggested `--dump` makes perfect sense to me.

**Comment by yanniszark** [comment]
Have we discussed this before? I'm having a deja-vu moment :sweat_smile: 
But yes, I agree this makes more sense with build.
`--dry-run` implies some sort of validation (and is what kubectl does)
I see this has been implemented as a separate function `Dump` so it should be a matter of changing a couple of files under `cmd` (unless I have missed something).

**Comment by discordianfish** [comment]
Yeah no disagreement from my side, just haven't gotten to it. Looking into that now.

**Comment by jlewi** [comment]
Looks like metadata test failed; probably unrelated to this PR. 
kubeflow/kfctl#57 is tracking getting rid of the overreliance on flaky E2E tests.

/test all

File: cmd/kfctl/cmd/build.go [hunk 1]
```diff
@@ -45,27 +45,30 @@ var buildCmd = &cobra.Command{
...
+				return fmt.Errorf("couldn't generate KfApp: %v", err)
 			}
-			return nil
-		} else {
+		default:
 			return fmt.Errorf("Unsupported object kind: %v", kind)
 		}
 
-		return err
+		if applyCfg.GetBool(string(kftypes.DUMP)) == true {
```
Comment (jlewi) [inline cmd/kfctl/cmd/build.go:68]: Should this be `buildCfg`  if we are binding build?

File: cmd/kfctl/cmd/build.go [hunk 2]
```diff
@@ -94,4 +97,13 @@ func init() {
 		log.Errorf("Couldn't set flag --%v: %v", string(kftypes.VERBOSE), bindErr)
 		return
 	}
+
+	// dump flag
+	applyCmd.Flags().BoolP(string(kftypes.DUMP), "d", false,
```
Comment (jlewi) [inline cmd/kfctl/cmd/build.go:102]: Should this be buildCmf and buildCfg if the command is build not apply?

**Comment by discordianfish** [comment]
Oops, yeah copy&paste mess. Fixed now. Also fixed the commit message.

**Comment by jlewi** [comment]
/lgmt
/approve

Thanks; unfortunately looks like there is merge conflict.

**Comment by jlewi** [comment]
THanks!
/lgtm
/approve

**Pull request merged**